src/compiler/scala/tools/nsc/typechecker/Contexts.scala

```diff
@@ -76,31 +70,55 @@ trait Contexts { self: Analyzer =>
 
   var lastAccessCheckDetails: String = ""
 
-  /** List of symbols to import from in a root context.  Typically that
-   *  is `java.lang`, `scala`, and [[scala.Predef]], in that order.  Exceptions:
+  lazy val generalRootImports: List[Import] =
+    settings.Ypredef.value
+      .map(global.newUnitParser(_, "<flag-Ypredef>").parseRule(_.importExpr()))
+      .collect { case imp: Import =>
+        // I can't seem to figure out how to use the freshly parsed tree...
+        // Instead we get to recreate it
+        gen.mkImportFromSelector(
+          rootMirror.getPackage(imp.expr.toString), imp.selectors)
+        }
+
+  /** List of imports for a unit's root context
    *
-   *  - if option `-Yno-imports` is given, nothing is imported
    *  - if the unit is java defined, only `java.lang` is imported
-   *  - if option `-Yno-predef` is given, if the unit body has an import of Predef
-   *    among its leading imports, or if the tree is [[scala.Predef]], `Predef` is not imported.
+   *  - otherwise, values from `-Ypredef` are imported.
+   *  - if a file explicitly imports a given predef, the value is not imported
+   *    an additional time
    */
-  protected def rootImports(unit: CompilationUnit): List[Symbol] = {
+  protected def rootImports(unit: CompilationUnit): List[Import] = {
     assert(definitions.isDefinitionsInitialized, "definitions uninitialized")
 
-    if (settings.noimports) Nil
-    else if (unit.isJava) RootImports.javaList
-    else if (settings.nopredef || treeInfo.noPredefImportForUnit(unit.body)) {
-      // SI-8258 Needed for the presentation compiler using -sourcepath, otherwise cycles can occur. See the commit
-      //         message for this ticket for an example.
-      debuglog("Omitted import of Predef._ for " + unit)
-      RootImports.javaAndScalaList
+    if (unit.isJava) gen.mkWildcardImport(JavaLangPackage) :: Nil
+    else {
+
+      /** Cleanup import `imp` such that any selectors covered by
+        * any imports in `parents` are removed.
+        */
+      def dedupe(imp: Import, parents: List[Import]): Option[Import] = {
```

Comment:
I am unsure of what the specific behavior should be for skipping predef imports. Also, what's the appropriate way to test this logic?